internal/cronsumer_test.go

```diff
@@ -0,0 +1,250 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/Trendyol/kafka-cronsumer/pkg/kafka"
+	"github.com/Trendyol/kafka-cronsumer/pkg/logger"
+	segmentio "github.com/segmentio/kafka-go"
+)
+
+func Test_Produce_Max_Retry_Count_Reach(t *testing.T) {
+	// Given
+	kafkaConfig := &kafka.Config{
+		Brokers: []string{"localhost:29092"},
+		Consumer: kafka.ConsumerConfig{
+			GroupID:  "sample-consumer",
+			Topic:    "exception",
+			Cron:     "@every 1s",
+			Duration: 20 * time.Second,
+		},
+		LogLevel: "info",
+		Logger:   logger.New("info"),
+	}
+
+	var firstConsumerFn kafka.ConsumeFn = func(message kafka.Message) error {
+		fmt.Printf("First consumer > Message received: %s\n", string(message.Value))
+		return nil
+	}
+	c := &kafkaCronsumer{
+		cfg:             kafkaConfig,
+		messageChannel:  make(chan MessageWrapper),
+		kafkaConsumer:   mockConsumer{},
+		kafkaProducer:   newProducer(kafkaConfig),
+		consumeFn:       firstConsumerFn,
+		metric:          &CronsumerMetric{},
+		maxRetry:        1,
+		deadLetterTopic: kafkaConfig.Consumer.DeadLetterTopic,
+	}
+	m := MessageWrapper{
+		Message: kafka.Message{
+			Headers: []kafka.Header{
+				{Key: RetryHeaderKey, Value: []byte("1")},
+			},
+			Topic: "exception",
+		},
+		RetryCount: 1,
+	}
+
+	// When
+	c.produce(m)
+
+	// Then
+	if !reflect.DeepEqual(c.metric.TotalDiscardedMessagesCounter, int64(1)) {
+		t.Errorf("Expected: %+v, Actual: %+v", c.metric.TotalDiscardedMessagesCounter, int64(1))
+	}
+}
+
+func Test_Produce_Max_Retry_Count_Reach_Dead_Letter_Topic_Feature_Enabled(t *testing.T) {
+	// Given
+	kafkaConfig := &kafka.Config{
+		Brokers: []string{"localhost:29092"},
+		Consumer: kafka.ConsumerConfig{
+			GroupID:  "sample-consumer",
+			Topic:    "exception",
+			Cron:     "@every 1s",
+			Duration: 20 * time.Second,
+		},
+		LogLevel: "info",
+		Logger:   logger.New("info"),
+	}
+
+	var firstConsumerFn kafka.ConsumeFn = func(message kafka.Message) error {
+		fmt.Printf("First consumer > Message received: %s\n", string(message.Value))
+		return nil
+	}
+	c := &kafkaCronsumer{
+		cfg:             kafkaConfig,
```

Comment:
it seems to initialize cfg (kafka.Config) is not necessary here. Because in order to enable dead letter topic you have to pass DeadLetterTopic field on RetryConfigurations, please remove and check its working or not bro